src/Http/Paginators/OffsetRequestPaginator.php

```diff
@@ -0,0 +1,184 @@
+// TODO 1: Look into serialising the Connector and original Request,
+//           to ensure that we can rebuild the paginator state without storing the entire multiverse.
+// TODO 2: Make it easier to extend the RequestPaginator. Preferably via callbacks, so we ideally don't even need separate classes.
+// TODO 3: Because both page-based and offset-based pagination just bumps the 'paging' number,
+//           would it make sense to do all pagination in _the_ RequestPaginator, but allow to specify the counter somehow?
+//         Maybe a callback that receives a copy of the original Request, as well as the latest Response (which has the corresponding latest Request and PendingRequest),
+//           and have that callback set the next 'page' on the new Request?
+
+/**
+ * @template TRequest of \Saloon\Contracts\Request
+ * @template TResponse of \Saloon\Contracts\Response
+ *
+ * @extends RequestPaginator<TRequest, TResponse>
+ */
+class OffsetRequestPaginator extends RequestPaginator
+{
+    protected readonly int $originalOffset;
+
+    /**
+     * @var  \Closure(\Saloon\Contracts\Response): bool;
+     */
+    protected readonly Closure $hasNextPage;
+
+    /**
+     * @param \Saloon\Contracts\Connector $connector
+     * @param \Saloon\Contracts\Request $originalRequest
+     * @param int<0, max>|null $limit
+     * @param callable(\Saloon\Contracts\Response): bool $hasNextPage
+     * @param int<0, max> $offset
+     */
+    public function __construct(
+        Connector $connector,
+        Request $originalRequest,
+        ?int $limit,
+        callable $hasNextPage,
+        protected int $offset = 0,
+    ) {
+        parent::__construct($connector, $originalRequest, $limit);
+
+        $this->originalOffset = $offset;
+        $this->hasNextPage = $hasNextPage(...);
+    }
+
+    /**
+     * @return int
+     */
+    public function page(): int
+    {
+        return $this->offset;
+    }
+
+    /**
+     * @return void
+     */
+    public function rewind(): void
+    {
+        parent::rewind();
+
+        $this->offset = 0;
+    }
+
+    /**
+     * @return bool
+     */
+    public function valid(): bool
+    {
+        // If we haven't sent a request yet, and therefore don't have any response, the iterator is still valid.
+        // It's only ever invalid when we have sent a request, retrieved the response, and have no more pages after that.
+        // Otherwise PHP would immediately terminate  the iteration, without sending a request.
+        if (\is_null($this->lastResponse)) {
+            return true;
+        }
+
+        return ($this->hasNextPage)($this->lastResponse);
+    }
+
+    /**
+     * @return ($this->async is true ? \GuzzleHttp\Promise\PromiseInterface : TResponse)
+     */
+    public function current(): Response|PromiseInterface
+    {
+        $request = clone $this->originalRequest;
+        $request->query()->merge([
+            'limit', $this->limit,
+            'offset', $this->offset,
+        ]);
```

Comment:
I wonder if it would be better to create a pending request and modify that instead, so the original request is left untouched:

```suggestion
        $pendingRequest = $this->connector->createPendingRequest(clone $this->originalRequest);
        $pendingRequest->query()->merge([
            'limit', $this->limit,
            'offset', $this->offset,
        ]);
```

Reply:
We never alter the original request - we clone it all the time.
This is to keep tabs on the literal initial request, which is the same regardless of amount of requests made.
Then we can follow the regular 'send' flow, by simply bumping pages.

I guess, technically speaking, the current and suggested ways are semantically equivalent.